File: ios/Sources/MessageBusPlugin/MessageBus.swift
final class MessageBus {
    static let shared = MessageBus()

    typealias Payload = Any?
    typealias Handler = (Payload) -> Void
    typealias GlobalHandler = (_ type: String, _ payload: Payload) -> Void

    final class Subscription {
        private let cancelClosure: () -> Void
        private var isActive = true
        init(cancel: @escaping () -> Void) { self.cancelClosure = cancel }
        public func cancel() { guard isActive else { return }; isActive = false; cancelClosure() }
        deinit { cancel() }
    }

    private var handlers: [String: [UUID: Handler]] = [:]
    private var globalHandlers: [UUID: GlobalHandler] = [:]
    private let queue = DispatchQueue(label: "MessageBus.queue", qos: .default)

    private init() {}

    @discardableResult
    func subscribe(type: String, _ handler: @escaping Handler) -> Subscription {
        let token = UUID()
        queue.sync {
            var bucket = handlers[type, default: [:]]
            bucket[token] = handler
            handlers[type] = bucket
        }
        return Subscription { [weak self] in self?.remove(type: type, token: token) }
    }

    @discardableResult
    func subscribeAll(_ handler: @escaping GlobalHandler) -> Subscription {
        let token = UUID()
        queue.sync { globalHandlers[token] = handler }
        return Subscription { [weak self] in self?.removeGlobal(token: token) }
    }

    func publish(type: String, payload: Payload) {
        var callbacks: [Handler] = []
        var globals: [GlobalHandler] = []
        queue.sync {
            callbacks = Array(handlers[type]?.values ?? [])
            globals = Array(globalHandlers.values)
        }
        callbacks.forEach { $0(payload) }
        globals.forEach { $0(type, payload) }
    }

    private func remove(type: String, token: UUID) {
        queue.sync {
            guard var bucket = handlers[type] else { return }
            bucket.removeValue(forKey: token)
            if bucket.isEmpty { handlers.removeValue(forKey: type) } else { handlers[type] = bucket }
        }
    }

    private func removeGlobal(token: UUID) {
        queue.sync { globalHandlers.removeValue(forKey: token) }
    }

    func removeAll(for type: String? = nil) {
        queue.sync {
            if let t = type { handlers.removeValue(forKey: t) }
            else { handlers.removeAll(); globalHandlers.removeAll() }
        }
    }
}

File: android/src/main/java/com/cnctd/plugins/messagebus/MessageBus.kt
package com.cnctd.plugins.messagebus

import java.util.UUID

class MessageBus private constructor() {
    companion object {
        val shared = MessageBus()
    }

    fun interface Handler {
        fun onMessage(payload: Any?)
    }

    fun interface GlobalHandler {
        fun onMessage(type: String, payload: Any?)
    }

    class Subscription(private val cancelAction: () -> Unit) {
        private var isActive = true

        @Synchronized
        fun cancel() {
            if (!isActive) return
            isActive = false
            cancelAction()
        }
    }

    private val lock = Any()
    private val handlers = mutableMapOf<String, MutableMap<String, Handler>>()
    private val globalHandlers = mutableMapOf<String, GlobalHandler>()

    fun subscribe(type: String, handler: Handler): Subscription {
        val token = UUID.randomUUID().toString()
        synchronized(lock) {
            handlers.getOrPut(type) { mutableMapOf() }[token] = handler
        }
        return Subscription { remove(type, token) }
    }

    fun subscribeAll(handler: GlobalHandler): Subscription {
        val token = UUID.randomUUID().toString()
        synchronized(lock) {
            globalHandlers[token] = handler
        }
        return Subscription { removeGlobal(token) }
    }

    fun publish(type: String, payload: Any? = null) {
        val callbacks: List<Handler>
        val globals: List<GlobalHandler>

        synchronized(lock) {
            callbacks = handlers[type]?.values?.toList() ?: emptyList()
            globals = globalHandlers.values.toList()
        }

        for (cb in callbacks) cb.onMessage(payload)
        for (g in globals) g.onMessage(type, payload)
    }

    private fun remove(type: String, token: String) {
        synchronized(lock) {
            val bucket = handlers[type] ?: return
            bucket.remove(token)
            if (bucket.isEmpty()) handlers.remove(type)
        }
    }

    private fun removeGlobal(token: String) {
        synchronized(lock) { globalHandlers.remove(token) }
    }

    fun removeAll(type: String? = null) {
        synchronized(lock) {
            if (type != null) {
                handlers.remove(type)
            } else {
                handlers.clear()
                globalHandlers.clear()
            }
        }
    }
}
